internal/processor/stream/processor.go

```diff
@@ -212,31 +218,70 @@ func (p *Processor) readBatch(
 // Callers must not access result concurrently with HandleStream.
 func (p *Processor) HandleStream(
 	ctx context.Context,
+	async bool,
 	baseEvent model.APMEvent,
-	reader io.Reader,
+	reader io.ReadCloser,
 	batchSize int,
 	processor model.BatchProcessor,
 	result *Result,
 ) error {
-	// Since processor.ProcessBatch can block for some time until all the batch
-	// is added to the modelindexer's cache and there isn't a fail-fast rejection,
-	// we want to cap how many in-flight requests are read at any time.
+	// Limits the number of concurrent batch decodes.
 	// Defaults to 200 (N), only allowing N requests to read and cache Y events
-	// (determined by batchSize) from the batch, effectively limitting the decoding
-	// concurrency to N batches at any time, which should be a good ceiling. The
-	// ceiling also reduces the contention on the modelindexer.activeMu.
-	select {
-	case p.sem <- struct{}{}:
-	case <-ctx.Done():
-		return ctx.Err()
+	// (determined by batchSize) from the batch.
+	// The ceiling may also reduce the contention on the modelindexer.activeMu.
+	// Clients can set a async to true which makes the processor handle the
+	// events in the background. Returns with an error `publish.ErrFull` if the
+	// semaphore is full.
+	if err := p.acquireLock(ctx, async); err != nil {
+		return err
 	}
 
 	sr := p.getStreamReader(reader)
-	defer func() {
+	releaseResources := func() {
 		sr.release()
-		<-p.sem
+		// Close the reader after the batch have been processed.
+		reader.Close()
+		p.releaseLock()
+	}
+
+	// By default, streams are handled synchronously, however, clients can
+	// set the request to be processed asynchronously, in which case, a
+	// the processing will happen in the background within an new goroutine,
+	// and since error cannot be sent back to the client, they are logged.
+	if !async {
+		defer releaseResources()
+		return p.handleStream(ctx, baseEvent, sr, batchSize, processor, result)
+	}
+	go func() {
```

Comment:
Instead of doing _everything_ async, I was thinking of just doing the processing (ProcessBatch) async.

So we would decode the body in the request handler into batches as before, and send them to a queue/channel. If the queue is full, we would return with a "queue is full" error. Otherwise, it goes on the queue to be picked up by another short lived goroutine to send to ProcessBatch.

Then the question becomes how to size the channel of course :)